File: 287-find-the-duplicate-number/find-the-duplicate-number.java
class Solution {
    public int findDuplicate(int[] nums) {
        int n=nums.length;
        int ans=-1;
        for(int i=0;i<n;i++){
            int index=Math.abs(nums[i]);
            if(nums[index]<0){
                ans=index;
                break;
            }
            nums[index]*=-1;
        }
        return ans;
    }
}

File: 287-find-the-duplicate-number/find-the-duplicate-number.cpp
class Solution {
public:
    int detectCycle(vector<int>& nums){
        int n = nums.size();
        int slowIdx = 0;
        int fastIdx = 0;
        while(true){
            slowIdx = nums[slowIdx];
            fastIdx = nums[nums[fastIdx]];
            if(slowIdx == fastIdx) break;
        }
        return slowIdx;
    }
    int findCycleStart(int meetIdx,vector<int>& nums){
        int p1 = 0;
        int p2 =meetIdx;

        while(p1!=p2){
            p1=nums[p1];
            p2=nums[p2];
        }
        return p2;
    }
    int findDuplicate(vector<int>& nums) {
        int meet = detectCycle(nums);
        return findCycleStart(meet,nums);
    }
};